### PR TITLE
First part of merge, private routes

### PR DESCRIPTION
**Pull request by Artin13**
First part of the merge. This is for the private/protected routes

### CODE REVIEW

File: client/src/index.tsx [hunk 1]
```diff
@@ -12,38 +12,68 @@ import DTHistory from './route/history/History';
...
 const router = createBrowserRouter([
   {
     path: '/',
     element: <SignIn />,
   },
   {
     path: 'dashboard',
-    element: <Dashboard />,
+    element: (
+      <PrivateRoute useDummyAuth={true}>
```
Comment (prasadtalasila) [inline client/src/index.tsx:25]: PrivateRoute must always authenticate. The option is not necessary.

File: script/install.bash [hunk 2]
```diff
@@ -15,27 +15,30 @@ apt-get install -y \
 
 mkdir -p /etc/apt/keyrings
 curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /etc/apt/keyrings/docker.gpg
-printf \
-  "deb [arch=%s signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu \
-  %s stable" "$(dpkg --print-architecture)" "$(lsb_release -cs)"  | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
+echo \
```
Comment (prasadtalasila) [inline script/install.bash:18]: This file has been linted in the previous two commits. You only have to make the following changes:

add the lines:
```
#install docker-compose from https://docs.docker.com/compose/install/other/
curl -SL "https://github.com/docker/compose/releases/download/v2.15.1/docker-compose-linux-x86_64" -o /usr/local/bin/docker-compose
ln -s /usr/local/bin/docker-compose /usr/bin/docker-compose
```
remove sudo from lines - 48, 52, 56 of the file in the latest commit on the `feature/distributed-demo` branch

File: client/src/components/PrivateRoute.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import * as React from 'react';
+import { Navigate } from 'react-router-dom';
+
+interface PrivateRouteProps {
+  children: React.ReactNode;
+  useDummyAuth?: boolean;
+}
+
+class DummyAuth {
```
Comment (prasadtalasila) [inline client/src/components/PrivateRoute.tsx:9]: Please add both unit and integration tests for this module

File: client/src/components/PrivateRoute.tsx [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+const PrivateRoute: React.FC<PrivateRouteProps> = ({ children, useDummyAuth = false }) => {
+  const [isAuthenticated, setIsAuthenticated] = React.useState<boolean | null>(null);
+
+  React.useEffect(() => {
+    const dummyAuth = new DummyAuth(useDummyAuth);
+    setIsAuthenticated(dummyAuth.checkAuth());
+  }, [useDummyAuth]);
+
+  if (isAuthenticated === null) {
+    return <div>Loading...</div>;
```
Comment (prasadtalasila) [inline client/src/components/PrivateRoute.tsx:30]: "User not signed in...please signin" (signin word will have a link to signin page)

File: client/src/components/PrivateRoute.tsx (hunk 3 above)
Comment (prasadtalasila) [reply client/src/components/PrivateRoute.tsx:9]: You can call this class as Auth. You don't need to rename later.

File: client/src/components/PrivateRoute.tsx [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  constructor(auth: boolean) {
+    this.isAuthenticated = auth;
+  }
+
+  checkAuth() {
+    return this.isAuthenticated;
+  }
+}
+
+const PrivateRoute: React.FC<PrivateRouteProps> = ({ children, useDummyAuth = false }) => {
```
Comment (prasadtalasila) [inline client/src/components/PrivateRoute.tsx:21]: this line is too long; you might want to insert newline after the first equals symbol.

**Review by prasadtalasila — CHANGES_REQUESTED** [review]

**Comment by prasadtalasila** [comment]
@KarstenMalle , I need to check the current best practice of React.FC and React.useState(). Otherwise, rest of the review is complete.

**Comment by prasadtalasila** [comment]
> @KarstenMalle , I need to check the current best practice of React.FC and React.useState(). Otherwise, rest of the review is complete.

This part looks cool. Please go ahead with the rest of the changes.

**Review by prasadtalasila — COMMENTED** [review]
already commented on the conversation thread. Gist is that the tests are failing.

**Comment by KarstenMalle** [comment]
> @KarstenMalle, thanks for the update. I've used the updated client website and the authentication functionality seems to be in place. However, most of the tests fail. Please see the Github Actions [log](https://github.com/INTO-CPS-Association/DTaaS/actions/runs/4722783711/jobs/8377843768?pr=40#step:6:36). Even `yarn test` shows the same error log.

Yes sorry about that. It should be fixed now.

**Comment by KarstenMalle** [comment]
> > @KarstenMalle, thanks for the update. I've used the updated client website and the authentication functionality seems to be in place. However, most of the tests fail. Please see the Github Actions [log](https://github.com/INTO-CPS-Association/DTaaS/actions/runs/4722783711/jobs/8377843768?pr=40#step:6:36). Even `yarn test` shows the same error log.
> 
> Yes sorry about that. I see the Playwright tests are also failing, I will try to fix it tonight.

Playwright tests are fixed, code climate is complaining about duplication but I think that is expected due to testing similar scenarios. Let me know if you want anything changed :)

**Pull request merged**